tsdb/head.go

```diff
@@ -1468,6 +1471,15 @@ func (h *Head) Delete(mint, maxt int64, ms ...*labels.Matcher) error {
 		h.tombstones.AddInterval(s.Ref, s.Intervals[0])
 	}
 
+	// New code for OOO data deletion
+	for _, oooChunk := range h.oooChunks {
+		oooChunk.DeleteRange(mint, maxt, oooChunk.seriesRef)
```

Comment:
We don't want to delete synchronously but rather make sure that the OOO head uses the tombstones.

When there is a delete call a new stone is added. This is the logic happening above. And you'll see that both the in order head and the out of order head implement Tombstones()
- https://github.com/prometheus/prometheus/blob/6ef9ed0bc33299499df9addee5b007070c5d32b8/tsdb/head.go#L1530-L1533
- https://github.com/prometheus/prometheus/blob/817a2396cb2b9d15c6e10d88fa3b6fffdc914d3e/tsdb/ooo_head.go#L132-L136

That is part of the BlockReader interface.


Then on queries, which happens either during compaction or a normal prometheus query, the tsdb treats blocks and both heads as just blocks and reads from them.
- During compaction https://github.com/prometheus/prometheus/blob/4451ba10b455a8b9c822d872639c9afa41e14abf/tsdb/compact.go#L727-L731
- During querying https://github.com/prometheus/prometheus/blob/69edd8709ba7d4b2152eab5fb208a3d74b4a81e1/tsdb/querier.go#L73-L78

The problem we have here is that the OOO head always returns an empty set of stones but we need to track the ones added during runtime.